src/elements/play-assets/play-assets.ts

```diff
@@ -0,0 +1,429 @@
+import {type PropertyValues, ReactiveElement} from 'lit'
+import {customElement, property} from 'lit/decorators.js'
+import {WebAccess, WebStorage} from '@zenfs/dom'
+import {Bubble} from '../../utils/bubble.js'
+import {Zip} from '@zenfs/zip'
+import {
+  hasFileAccessAPI,
+  tryGetFile,
+  tryQueryPermission
+} from '../../utils/file-access-api.js'
+import * as ZenFS from '@zenfs/core'
+import {FileSystem, fs, InMemory} from '@zenfs/core'
+import type {AssetMap} from '@devvit/shared-types/Assets.js'
+import * as IDB from 'idb-keyval'
+
+declare global {
+  interface HTMLElementEventMap {}
+  interface HTMLElementTagNameMap {
+    'play-assets': PlayAssets
+  }
+}
+
+export type AssetFilesystemType = 'virtual' | 'local'
+
+const CACHE_LAST_DIR = 'lastMountedDirectory'
+const CACHE_LAST_ZIP = 'lastMountedArchive'
+
+const MIME: {[ext: string]: string} = {
+  html: 'text/html',
+  htm: 'text/html',
+  jpg: 'image/jpeg',
+  jpeg: 'image/jpeg',
+  png: 'image/png',
+  gif: 'image/gif'
+}
+const DEFAULT_MIME = 'application/octet-stream'
+
+declare global {
+  interface HTMLElementTagNameMap {
+    'play-assets': PlayAssets
+  }
+  interface HTMLElementEventMap {
+    'assets-updated': CustomEvent<void>
+  }
+}
+
+@customElement('play-assets')
+export class PlayAssets extends ReactiveElement {
+  @property({attribute: 'allow-storage', type: Boolean})
+  allowStorage: boolean = false
+
+  @property({attribute: 'filesystem-type', reflect: true, type: String})
+  filesystemType: AssetFilesystemType = 'virtual'
+
+  @property({attribute: false})
+  directoryHandle: FileSystemDirectoryHandle | undefined
+
+  @property({attribute: false})
+  directoryName: string | undefined
+
+  @property({attribute: false})
+  archiveHandle: FileSystemFileHandle | File | undefined
+
+  @property({attribute: false})
+  archiveFilename: string | undefined
+
+  @property({attribute: false})
+  assetCount: number = 0
+
+  _rootAssets: PlayAssets | undefined
+  #assetMap: AssetMap = {}
+
+  private _waitForInit: Promise<void> = new Promise(
+    resolve => (this._initComplete = resolve)
+  )
+  private _initComplete!: () => void
+
+  static get hasFileAccessAPI(): boolean {
+    return hasFileAccessAPI
+  }
+
+  get assetMap(): Promise<AssetMap> {
+    if (this._rootAssets) {
+      return this._rootAssets.assetMap
+    }
+
+    const resolvedMap = () => this._waitForInit.then(() => this.#assetMap)
```

Comment:
can we make this async to thenable syntax? does updateComplete suffice instead of _waitForInit()?